pkg/registry/certificates/strategy.go

```diff
@@ -0,0 +1,207 @@
+	csr := obj.(*certificates.CertificateSigningRequest)
+	csr.Spec.Fingerprint = ""
+	csr.Spec.Subject = certificates.Subject{}
+	csr.Spec.Hostnames = []string{}
+	csr.Spec.IPAddresses = []string{}
+
+	// Be explicit that users cannot create pre-approved certificate requests.
+	csr.Status = certificates.CertificateSigningRequestStatus{}
+	csr.Status.Conditions = []certificates.CertificateSigningRequestCondition{}
+}
+
+// PrepareForUpdate clears fields that are not allowed to be set by end users
+// on update. Certificate requests are immutable after creation except via subresources.
+func (csrStrategy) PrepareForUpdate(obj, old runtime.Object) {
+	newCSR := obj.(*certificates.CertificateSigningRequest)
+	oldCSR := old.(*certificates.CertificateSigningRequest)
+
+	newCSR.Spec = oldCSR.Spec
+	newCSR.Status = oldCSR.Status
+}
+
+// Validate validates a new CSR. Validation must check for a correct signature.
+func (csrStrategy) Validate(ctx api.Context, obj runtime.Object) field.ErrorList {
+	csr := obj.(*certificates.CertificateSigningRequest)
+	return validation.ValidateCertificateSigningRequest(csr)
+}
+
+// Canonicalize normalizes the object after validation (which includes a
+// signature check). This ordering makes it an ideal place to extract some
+// trusted CSR information.
+func (csrStrategy) Canonicalize(obj runtime.Object) {
+	csr := obj.(*certificates.CertificateSigningRequest)
+
+	// TODO: assuming this all works because validation just did it
+	pemBytes, _ := base64.StdEncoding.DecodeString(csr.Spec.Request)
+	block, _ := pem.Decode(pemBytes)
+	request, _ := x509.ParseCertificateRequest(block.Bytes)
+
+	// internal.Subject is just a marshalling wrapper around pkix.Name
+	csr.Spec.Subject = certificates.NewInternalSubject(request.Subject)
+	csr.Spec.Hostnames = request.DNSNames
+	for _, ip := range request.IPAddresses {
+		csr.Spec.IPAddresses = append(csr.Spec.IPAddresses, ip.String())
+	}
+	fingerprint, err := calculateFingerprint(request)
+	if err != nil {
+		glog.Errorf("failed to calculate fingerprint of CSR %s: %v", csr.Name, err)
+	} else {
+		csr.Spec.Fingerprint = fingerprint
+	}
+}
+
+// ValidateUpdate is the default update validation for an end user.
+func (csrStrategy) ValidateUpdate(ctx api.Context, obj, old runtime.Object) field.ErrorList {
+	errorList := validation.ValidateCertificateSigningRequest(obj.(*certificates.CertificateSigningRequest))
+	return append(errorList, validation.ValidateCertificateSigningRequestUpdate(obj.(*certificates.CertificateSigningRequest), old.(*certificates.CertificateSigningRequest))...)
+}
+
+// If AllowUnconditionalUpdate() is true and the object specified by
+// the user does not have a resource version, then generic Update()
+// populates it with the latest version. Else, it checks that the
+// version specified by the user matches the version of latest etcd
+// object.
+func (csrStrategy) AllowUnconditionalUpdate() bool {
+	return true
```

Comment:
At one point I was trying to make it easier to manage requests with third-party clients (read: curl/jq/generic scripting languages). Is it bad practice to have this be true?